### PR TITLE
Add support for odfe roles

### PR DESCRIPTION
**Pull request by tdegiacinto**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -10,7 +10,7 @@ env:
   matrix:
     - ES_VERSION=5.6.16 ES_OSS_IMAGE=elasticsearch:${ES_VERSION} ES_IMAGE=docker.elastic.co/elasticsearch/elasticsearch:${ES_VERSION} ES_COMMAND="elasticsearch -Epath.repo=/tmp"
     - ES_VERSION=6.8.0 ES_OSS_IMAGE=docker.elastic.co/elasticsearch/elasticsearch-oss:${ES_VERSION} ES_IMAGE=docker.elastic.co/elasticsearch/elasticsearch:${ES_VERSION} ES_OPENDISTRO_IMAGE=amazon/opendistro-for-elasticsearch:0.9.0 XPACK=1
-    - ES_VERSION=7.0.1 ES_OSS_IMAGE=docker.elastic.co/elasticsearch/elasticsearch-oss:${ES_VERSION} ES_IMAGE=docker.elastic.co/elasticsearch/elasticsearch:${ES_VERSION} XPACK=1
+    - ES_VERSION=7.0.1 ES_OSS_IMAGE=docker.elastic.co/elasticsearch/elasticsearch-oss:${ES_VERSION} ES_IMAGE=docker.elastic.co/elasticsearch/elasticsearch:${ES_VERSION} ES_OPENDISTRO_IMAGE=amazon/opendistro-for-elasticsearch:1.2.0 XPACK=1
```
Comment (JHeilCoveo) [inline .travis.yml:13]: You should use ES_VERSION 7.2.x with ODFE 1.2.0

File: resource_elasticsearch_odfe_role_test.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+		allowed = true
+	}
+
+	randomName := "test" + acctest.RandStringFromCharSet(10, acctest.CharSetAlpha)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			if !allowed {
+				t.Skip("Roles only supported on ES >= 6")
```
Comment (JHeilCoveo) [inline resource_elasticsearch_odfe_role_test.go]: This comment seems to not be coherent with the previous switch case

**Review by JHeilCoveo — APPROVED** [review]

**Review by ntremblier — APPROVED** [review]

File: .travis.yml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 sudo: required
 language: go
 go:
-  - '1.12'
+  - "1.12"
```
Comment (jocgir) [inline .travis.yml]: Why did you change single quote for double quote?

File: resource_elasticsearch_odfe_role.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+}
+
+func resourceElasticsearchOdfeRoleDelete(d *schema.ResourceData, m interface{}) error {
+	var err error
+
+	path, err := uritemplates.Expand("/_opendistro/_security/api/roles/{name}", map[string]string{
+		"name": d.Get("role_name").(string),
+	})
+	if err != nil {
+		return fmt.Errorf("error building URL path for role: %+v", err)
```
Comment (jocgir) [inline resource_elasticsearch_odfe_role.go:148]: Strange to use `%+v` here. You should upgrade to go 1.13 and use `%w` instead.

File: resource_elasticsearch_odfe_role.go [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+	d.Set("tenant_permissions", res.TenantPermissions)
+	d.Set("cluster_permissions", res.ClusterPermissions)
+	d.Set("index_permissions", res.IndexPermissions)
+	d.Set("description", res.Description)
+
+	return nil
+}
+
+func resourceElasticsearchOdfeRoleUpdate(d *schema.ResourceData, m interface{}) error {
+	_, err := resourceElasticsearchPutOdfeRole(d, m)
```
Comment (jocgir) [inline resource_elasticsearch_odfe_role.go]: There is no use to split this into 2 statements. err is not needed after the if.
```suggestion
	if _, err := resourceElasticsearchPutOdfeRole(d, m); err != nil {
	    return err
	}
```

File: resource_elasticsearch_odfe_role.go [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+	}
+
+	rolesDefinition := RoleBody{
+		ClusterPermissions: expandStringList(d.Get("cluster_permissions").(*schema.Set).List()),
+		IndexPermissions:   indexPermissionsBody,
+		TenantPermissions:  tenantPermissionsBody,
+		Description:        d.Get("description").(string),
+	}
+	roleJSON, err := json.Marshal(rolesDefinition)
+
```
Comment (jocgir) [inline resource_elasticsearch_odfe_role.go:242]: Remove the blank line here, the if clearly goes with the assignation.

File: resource_elasticsearch_odfe_role.go [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+	if err != nil {
+		return response, fmt.Errorf("Body Error : %s", roleJSON)
+	}
+
+	path, err := uritemplates.Expand("/_opendistro/_security/api/roles/{name}", map[string]string{
+		"name": d.Get("role_name").(string),
+	})
+
```
Comment (jocgir) [inline resource_elasticsearch_odfe_role.go]: same

File: resource_elasticsearch_odfe_roles_mapping.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+	if err != nil {
+		return err
+	}
+
+	return resourceElasticsearchOdfeRolesMappingRead(d, m)
+}
+
+func resourceElasticsearchOdfeRolesMappingDelete(d *schema.ResourceData, m interface{}) error {
+	var err error
```
Comment (jocgir) [inline resource_elasticsearch_odfe_roles_mapping.go]: Why declaring err here?

File: resource_elasticsearch_odfe_roles_mapping.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+	d.Set("hosts", res.Hosts)
+	d.Set("users", res.Users)
+	d.Set("description", res.Description)
+	d.Set("and_backend_roles", res.AndBackendRoles)
+
+	return nil
+}
+
+func resourceElasticsearchOdfeRolesMappingUpdate(d *schema.ResourceData, m interface{}) error {
+	_, err := resourceElasticsearchPutOdfeRolesMapping(d, m)
```
Comment (jocgir) [inline resource_elasticsearch_odfe_roles_mapping.go]: Combine

File: resource_elasticsearch_odfe_role.go (hunk 4 above)
Comment (jocgir) [reply resource_elasticsearch_odfe_role.go:148]: This comment is applicable everywhere.

**Review by jocgir — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 1 above)
Comment (tdegiacinto) [reply .travis.yml:13]: This setting isn't related to my change I prefer to not change behaviour on something I didn't modify

File: .travis.yml (hunk 3 above)
Comment (tdegiacinto) [reply .travis.yml]: Linter choice

File: resource_elasticsearch_odfe_role.go (hunk 4 above)
Comment (tdegiacinto) [reply resource_elasticsearch_odfe_role.go:148]: we are not owner of this provider so I won't take this charge

File: .travis.yml (hunk 3 above)
Comment (jocgir) [reply .travis.yml]: Then, it should be `1.13.x`
no single or double quote

File: resource_elasticsearch_odfe_role.go (hunk 4 above)
Comment (jocgir) [reply resource_elasticsearch_odfe_role.go:148]: Do we push back our change to the upstream?

File: resource_elasticsearch_odfe_roles_mapping.go (hunk 8 above)
Comment (tdegiacinto) [reply resource_elasticsearch_odfe_roles_mapping.go]: leftover. Removed

File: resource_elasticsearch_odfe_role.go (hunk 4 above)
Comment (tdegiacinto) [reply resource_elasticsearch_odfe_role.go:148]: yes

File: resource_elasticsearch_odfe_role.go [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+		return err
+	}
+
+	name := d.Get("role_name").(string)
+	d.SetId(name)
+	return resourceElasticsearchOdfeRoleRead(d, m)
+}
+
+func resourceElasticsearchOdfeRoleRead(d *schema.ResourceData, m interface{}) error {
+	res, err := resourceElasticsearchGetOdfeRole(d.Id(), m)
```
Comment (jocgir) [inline resource_elasticsearch_odfe_role.go:114]: Here, I would rewrite like this:
```suggestion
	res, err := resourceElasticsearchGetOdfeRole(d.Id(), m)
	if err != nil {
		if elastic7.IsNotFound(err) {
			log.Printf("[WARN] OdfeRole (%s) not found, removing from state", d.Id())
			d.SetId("")
			return nil
		}
		return err
	}
```
It is weird to call `elastic7.IsNotFound` if there is no error.

**Review by jocgir — APPROVED** [review]

**Review by jocgir — APPROVED** [review]

**Review by mouellet-coveo — APPROVED** [review]

**Review by JHeilCoveo — APPROVED** [review]
